### PR TITLE
[Bug fix] Correct DeepSeek post-FF2 reduce-scatter reference

### PR DESCRIPTION
**Pull request by nkapreTT**
## Summary

- Construct the post-FF2 reduce-scatter input at the production width on every Galaxy device instead of sharding one device-width tensor across the mesh.
- Build and validate all 32 expected output shards at the production PCC threshold of `0.9999`.
- Explicitly skip prefill lengths above 128 because materializing the unchunked production-width fixture would require impractical host memory; decode and prefill-128 coverage remain enabled.
- Correct the single-shard reference documentation.

The previous test exercised an artificial `896 -> 112` composite edge case. Production DeepSeek emits a 7168-wide FF2 partial on each column and reduces it to a tile-aligned 896-wide shard.

## Validation

- exact PR head `461bb1bc73059a031be77ab1da11edeec0687245` on validated main base `fa1eb11ad481cf1465e7b716415d7a2b9d496e8c`: 26 successful checks, 50 skipped, zero pending/failures;
- Python compilation and the full changed-file pre-commit suite pass on the exact PR head;
- exact current-head Galaxy qualification [32395851318](https://github.com/tenstorrent/craq-sim/actions/runs/32395851318):
  - production-width `ds_reduce_scatter_post_ff2` PASS, `gsim_exit=0`;
  - all 32 output shards passed PCC, minimum `0.99999177022`, last `0.9999918310330201`;
  - pytest `1166.64s`, suite `1178s`, no errors;
- A/B Galaxy qualification [32387600969](https://github.com/tenstorrent/craq-sim/actions/runs/32387600969) established the regression: corrected reference minimum PCC `0.99999177022`; the prior reference failed at `0.872422042174`;
- stable aggregate patch-id: `777eee909318f1b09f41bf4f98a1c41360405f9f`.

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:codex/deepseek-rs-reference-20260820)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:codex/deepseek-rs-reference-20260820)

### CODE REVIEW

File: models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py [hunk 1]
```diff
@@ -349,17 +374,18 @@ def _build_reduce_scatter_inputs(
     mesh_width = mesh_device.shape[1]
     dim = hf_config.hidden_size
 
-    # Create input tensor with shape [num_layers, ..., dim]
-    # This will be sharded: each device gets [1, ..., dim/mesh_width] after sharding on dims=(0, -1)
+    # Create input tensor with shape [num_layers, ..., dim * mesh_width].
+    # This matches the post-FF2 production path: every column device owns a
+    # full-width partial product of size `dim`, and reduce-scatter sums those
+    # partial products before scattering the reduced result across columns.
+    input_width = dim * mesh_width
```
[the remark this answers is not in the document]
Comment (nkapreTT) [reply models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py:443]: Addressed by skipping the unchunked production-width prefill fixtures above sequence length 128 in both functional and device-performance paths. This avoids the 14 GiB through 1.75 TiB host allocations while retaining decode and prefill-128 production-shape coverage. Exact current-head Galaxy run 32395851318 passed all 32 shards (minimum PCC 0.99999177022, gsim_exit=0).

File: models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py [hunk 2]
```diff
@@ -81,23 +81,35 @@ def ds_reduce_scatter_post_ff2_reference(
         Shape is [num_layers, seq_len, batch, dim / mesh_width] for decode
         or [num_layers, batch, seq_len, dim / mesh_width] for prefill.
     """
-    # Get the dimensions
-    last_dim = x.shape[-1]
-    per_device_input_dim = last_dim // mesh_width  # dim - what each device has as input
-    per_device_output_dim = per_device_input_dim // mesh_width  # dim / mesh_width - output per device
+    return _build_reduce_scatter_post_ff2_reference_shards(x, mesh_width)[0]
```
[the remark this answers is not in the document]
Comment (nkapreTT) [reply models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py:88]: Addressed: the return contract now documents a leading dimension of 1 and explicitly identifies the row-0/column-0 shard. Exact current-head Galaxy run 32395851318 passed all 32 output shards at the production threshold.

**Comment by yieldthought** [comment]
Codex review:

[models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py:209](https://github.com/tenstorrent/tt-metal/blob/461bb1bc73059a031be77ab1da11edeec0687245/models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py#L209) assumes `ttnn.get_device_tensors()` returns every shard in the global mesh and that the returned list starts at row 0. In DUAL/QUAD distributed runs it exposes only the rank-local shards, while `ref_outputs` covers the full 8x8/16x8 mesh, so this assertion fails; without the assertion, later ranks would still be compared against row-0 references through `device_idx`. The file explicitly accepts DUAL/QUAD at [models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py:573](https://github.com/tenstorrent/tt-metal/blob/461bb1bc73059a031be77ab1da11edeec0687245/models/demos/deepseek_v3/tests/fused_op_unit_tests/mlp/test_ds_reduce_scatter_post_ff2.py#L573), and [models/demos/deepseek_v3/tests/unit/test_all_gather.py:97](https://github.com/tenstorrent/tt-metal/blob/461bb1bc73059a031be77ab1da11edeec0687245/models/demos/deepseek_v3/tests/unit/test_all_gather.py#L97) shows the established local-coordinate handling.

Suggestion: build the coordinate iterator from `tt_output.tensor_topology().mesh_coords()`, filter it through `mesh_device.get_view().is_local()` when distributed, assert against the local tensor count, and select `ref_outputs[row * mesh_width + col]` from each actual coordinate.

**Comment by nkapreTT** [comment]
Review-ready at 1be64de509: all review threads are resolved, with 26 successful checks and zero pending or failing checks. The pending CODEOWNER reviewers are already requested.

**Review by yieldthought — APPROVED** [review]
